src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -339,21 +325,31 @@ internal TypeSymbolWithAnnotations MergeNullability(TypeSymbolWithAnnotations ot
                 {
                     return null;
                 }
-                if (IsAnnotated)
-                {
-                    return true;
-                }
-                if (NonNullTypesContext.NonNullTypes == true)
-                {
-                    return _treatPossiblyNullableReferenceTypeTypeParameterAsNullable &&
-                        TypeSymbol.IsPossiblyNullableReferenceTypeTypeParameter();
-                }
-                if (TypeSymbol.IsValueType)
+
+                switch (NullableAnnotation)
                 {
-                    Debug.Assert(!TypeSymbol.IsNullableType());
-                    return false;
+                    case NullableAnnotation.Unknown:
+                        Debug.Assert(!TypeSymbol.IsNullableType());
+                        if (TypeSymbol.IsValueType)
+                        {
+                            return false;
+                        }
+
+                        return null;
+
+                    case NullableAnnotation.Nullable:
+                    case NullableAnnotation.NullableBasedOnAnalysis:
+                        return true;
+
+                    case NullableAnnotation.NotNullableBasedOnAnalysis:
+                        return false;
```

Comment:
>return false; [](start = 24, length = 13)

nit: consider collapsing the two "not nullable" cases, as you did for the two "nullable" cases.
We could probably also assert in "not nullable" cases, as you did for the "unknown" case (`Debug.Assert(!TypeSymbol.IsNullableType());`).